apps/system/flash_test/flash_test.c

```diff
@@ -85,16 +101,53 @@ static int flash_read(unsigned long addr, int size)
 	FAR uint8_t *buffer;
 	FAR struct mtd_dev_s *dev_mtd = NULL;
 
+	long long total_duration_ms = 0;
+
 	if (addr == 0 || size <= 0) {
 		printf("Invalid arguments, offset 0x%x, read size %d\n", addr, size);
 		return ERROR;
 	}
+	if (flash == FLASH_INTERNAL) {
+		printf("Received Flash Read cmd for internal flash offset 0x%x, read size %d\n", addr, size);
+		dev_mtd = up_flashinitialize();
+		if (!dev_mtd) {
+			printf("up_flashinitialize Failed\n");
+			return ERROR;
+		}
 
-	dev_mtd = up_flashinitialize();
-	if (!dev_mtd) {
-		printf("up_flashinitialize Failed\n");
-		return ERROR;
 	}
+#ifdef CONFIG_SECOND_FLASH_PARTITION
+
+	else if (flash == FLASH_EXT) {
+		printf("Received Flash Read cmd for external flash offset 0x%x, read size %d\n", addr, size);
+		struct spi_dev_s *spi = up_spiinitialize(1);
+
+#ifdef CONFIG_MTD_JEDEC
+		dev_mtd = jedec_initialize(spi);
+		if (dev_mtd == NULL) {
+			lldbg("Jedec Init failed\n");
+			return ERROR;
+		}
+#elif defined(CONFIG_MTD_W25)
+		dev_mtd = w25_initialize(spi);
+		if (dev_mtd == NULL) {
+			lldbg("w25 Init failed\n");
+			return ERROR;
+		}
+		printf("w25_initialize: done for secondary flash \n");
+#endif
```

Comment:
This kind of code should be done in kernel, not here. And they are done already. 
As I left a comment before, let's use bch to read / write some data from/to a flash.

Reply:
Direct flash access used for flat config only. update the app accordingly.